src/adapters/sqlite/encodeBatch/index.js

```diff
@@ -0,0 +1,129 @@
+// @flow
+
+import type { RecordId } from '../../../Model'
+import type { TableName, TableSchema, AppSchema } from '../../../Schema'
+import type { RawRecord } from '../../../RawRecord'
+import type { BatchOperation } from '../../type'
+import { validateTable } from '../../common'
+import type { SQL, SQLiteArg, NativeBridgeBatchOperation } from '../type'
+
+export function encodeInsertSql(schema: TableSchema): SQL {
+  const columns = schema.columnArray
+  const columnsSql = `"id", "_status", "_changed${columns
+    .map((column) => `", "${column.name}`)
+    .join('')}"`
+  const placeholders = Array(columns.length + 3)
+    .fill('?')
+    .join(', ')
+  return `insert into "${schema.name}" (${columnsSql}) values (${placeholders})`
+}
+
+export function encodeInsertArgs(tableSchema: TableSchema, raw: RawRecord): SQLiteArg[] {
+  const columns = tableSchema.columnArray
+  const len = columns.length
+
+  const args = Array(len + 3)
+  args[0] = raw.id
+  args[1] = raw._status
+  args[2] = raw._changed
+  for (let i = 0; i < len; i++) {
+    args[i + 3] = raw[(columns[i].name: any)]
+  }
+
+  return args
+}
+
+export function encodeUpdateSql(schema: TableSchema): SQL {
+  const columns = schema.columnArray
+  const placeholders = columns.map((column) => `, "${column.name}" = ?`).join('')
+  return `update "${schema.name}" set "_status" = ?, "_changed" = ?${placeholders} where "id" is ?`
+}
+
+export function encodeUpdateArgs(tableSchema: TableSchema, raw: RawRecord): SQLiteArg[] {
+  const columns = tableSchema.columnArray
+  const len = columns.length
+
+  const args = Array(len + 3)
+  args[0] = raw._status
+  args[1] = raw._changed
+  for (let i = 0; i < len; i++) {
+    args[i + 2] = raw[(columns[i].name: any)]
+  }
+  args[len + 2] = raw.id
+
+  return args
+}
+
+type GrouppedBatchOperation =
+  | ['create', TableName<any>, RawRecord[]]
+  | ['update', TableName<any>, RawRecord[]]
+  | ['markAsDeleted', TableName<any>, RecordId[]]
+  | ['destroyPermanently', TableName<any>, RecordId[]]
+
+export function groupOperations(operations: BatchOperation[]): GrouppedBatchOperation[] {
+  const grouppedOperations: GrouppedBatchOperation[] = []
+  let previousType: ?string = null
+  let previousTable: ?TableName<any> = null
+  let currentOperation: ?GrouppedBatchOperation = null
+  operations.forEach((operation) => {
+    const [type, table, rawOrId] = operation
+    if (type !== previousType || table !== previousTable) {
+      if (currentOperation) {
+        grouppedOperations.push(currentOperation)
+      }
+      previousType = type
+      previousTable = table
+      // $FlowFixMe
+      currentOperation = [type, table, []]
+    }
+
+    // $FlowFixMe
+    currentOperation[2].push(rawOrId)
+  })
+  if (currentOperation) {
+    grouppedOperations.push(currentOperation)
+  }
+  return grouppedOperations
+}
```

Comment:
IMO reducing `operations` with a `[groupedOps, currentOp]` accumulator could be a bit more cleaner

Reply:
probably, but this is performance-critical

Reply:
fixed in #1040